{
  "issue_number": 145.0,
  "title": "Server crash on failed master/slave sync (refused connection)",
  "body": "Three of my instances (on server 1 and 2) just crashed when I restarted server 3. Instance A was a master to an instance on server 3. Instances B and C were slaves of two instances on server 3.\n\nI'm sorry that I have to be so vague about this. I just found these stackdumps in my logfiles a few minutes after it happened. I believe, instances A,B,C were trying to do some kind of syncing with server 3, which was being restarted at the time. I assume server 3 refused the connections when the instances were in a state where these exceptions weren't handled.\n\nI'm using the fresh 2.4.1 from this morning on Ubuntu Oneiric. Let me know if you need more input.\n\ninstance 1:\n\n```\n[2710] 17 Oct 16:56:25 * MASTER MODE enabled (user request)\n[2710] 17 Oct 16:56:25 * Slave ask for synchronization\n[2710] 17 Oct 16:56:25 * Starting BGSAVE for SYNC\n[2710] 17 Oct 16:56:25 * Background saving started by pid 4019\n[2710] 17 Oct 16:56:27 - DB 0: 91 keys (0 volatile) in 128 slots HT.\n[2710] 17 Oct 16:56:27 - 0 clients connected (1 slaves), 92076368 bytes in use\n[2710] 17 Oct 16:56:30 * Non blocking connect for SYNC fired the event.\n[2710] 17 Oct 16:56:30 # I/O error writing to MASTER: Connection refused\n[2710] 17 Oct 16:56:30 * Connecting to MASTER...\n[2710] 17 Oct 16:56:30 # ======= Ooops! Redis 2.4.1 got signal: -11- =======\n[2710] 17 Oct 16:56:30 # redis_version:2.4.1\nredis_git_sha1:ab648d97\nredis_git_dirty:0\narch_bits:64\nmultiplexing_api:epoll\nprocess_id:2710\nuptime_in_seconds:4920\nuptime_in_days:0\nlru_clock:1863635\nused_cpu_sys:16.91\nused_cpu_user:6.16\nused_cpu_sys_children:0.00\nused_cpu_user_children:0.00\nconnected_clients:0\nconnected_slaves:1\nclient_longest_output_list:0\nclient_biggest_input_buf:0\nblocked_clients:0\nused_memory:92076368\nused_memory_human:87.81M\nused_memory_rss:96280576\nused_memory_peak:1022465704\nused_memory_peak_human:975.10M\nmem_fragmentation_ratio:1.05\nmem_allocator:jemalloc-2.2.1\nloading:0\naof_enabled:1\nchanges_since_last_save:16060\nbgsave_in_progress:1\nlast_save_time:1318865670\nbgrewriteaof_in_progress:0\ntotal_connections_received:2475\ntotal_commands_processed:13985\nexpired_keys:0\nevicted_keys:0\nkeyspace_hits:1204\nkeyspace_misses:9139\npubsub_channels:0\npubsub_patterns:0\nlatest_fork_usec:28578\nvm_enabled:0\nrole:master\naof_current_size:91031534\naof_base_size:0\naof_pending_rewrite:0\ndb0\n[2710] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f3fc4b64215]\n[2710] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f3fc4b64215]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server() [0x40ca99]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(connectWithMaster+0x1a) [0x41c14a]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(replicationCron+0xe6) [0x41c416]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(serverCron+0x1c8) [0x40fa78]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(aeProcessEvents+0x1db) [0x40c45b]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(aeMain+0x2e) [0x40c70e]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server(main+0x106) [0x40b606]\n[2710] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed) [0x7f3fc4a9530d]\n[2710] 17 Oct 16:56:30 # /usr/local/bin/redis-server() [0x40b72d]\n[4019] 17 Oct 16:56:32 * DB saved on disk\n```\n\ninstance 2:\n\n```\n[2698] 17 Oct 16:56:30 * Non blocking connect for SYNC fired the event.\n[2698] 17 Oct 16:56:30 # I/O error writing to MASTER: Connection refused\n[2698] 17 Oct 16:56:30 * Connecting to MASTER...\n[2698] 17 Oct 16:56:30 # ======= Ooops! Redis 2.4.1 got signal: -11- =======\n[2698] 17 Oct 16:56:30 # redis_version:2.4.1\nredis_git_sha1:ab648d97\nredis_git_dirty:0\narch_bits:64\nmultiplexing_api:epoll\nprocess_id:2698\nuptime_in_seconds:4920\nuptime_in_days:0\nlru_clock:1863635\nused_cpu_sys:17.05\nused_cpu_user:6.90\nused_cpu_sys_children:0.14\nused_cpu_user_children:0.62\nconnected_clients:0\nconnected_slaves:1\nclient_longest_output_list:0\nclient_biggest_input_buf:0\nblocked_clients:0\nused_memory:87061824\nused_memory_human:83.03M\nused_memory_rss:92336128\nused_memory_peak:1018799560\nused_memory_peak_human:971.60M\nmem_fragmentation_ratio:1.06\nmem_allocator:jemalloc-2.2.1\nloading:0\naof_enabled:1\nchanges_since_last_save:0\nbgsave_in_progress:0\nlast_save_time:1318870572\nbgrewriteaof_in_progress:0\ntotal_connections_received:2565\ntotal_commands_processed:13986\nexpired_keys:0\nevicted_keys:0\nkeyspace_hits:1034\nkeyspace_misses:8965\npubsub_channels:0\npubsub_patterns:0\nlatest_fork_usec:26442\nvm_enabled:0\nrole:master\naof_current_size:86031167\naof_base_size:0\naof_pending_rewrite:0\ndb0:key\n[2698] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f402ca77215]\n[2698] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f402ca77215]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server() [0x40ca99]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(connectWithMaster+0x1a) [0x41c14a]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(replicationCron+0xe6) [0x41c416]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(serverCron+0x1c8) [0x40fa78]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(aeProcessEvents+0x1db) [0x40c45b]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(aeMain+0x2e) [0x40c70e]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server(main+0x106) [0x40b606]\n[2698] 17 Oct 16:56:30 # /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed) [0x7f402c9a830d]\n[2698] 17 Oct 16:56:30 # /usr/local/bin/redis-server() [0x40b72d]\n```\n\ninstance 3:\n\n```\n[2684] 17 Oct 16:56:31 * Non blocking connect for SYNC fired the event.\n[2684] 17 Oct 16:56:31 # I/O error writing to MASTER: Connection refused\n[2684] 17 Oct 16:56:31 * Connecting to MASTER...\n[2684] 17 Oct 16:56:31 # ======= Ooops! Redis 2.4.1 got signal: -11- =======\n[2684] 17 Oct 16:56:31 # redis_version:2.4.1\nredis_git_sha1:ab648d97\nredis_git_dirty:0\narch_bits:64\nmultiplexing_api:epoll\nprocess_id:2684\nuptime_in_seconds:4916\nuptime_in_days:0\nlru_clock:1863635\nused_cpu_sys:28.23\nused_cpu_user:8.24\nused_cpu_sys_children:1.48\nused_cpu_user_children:6.89\nconnected_clients:0\nconnected_slaves:1\nclient_longest_output_list:0\nclient_biggest_input_buf:0\nblocked_clients:0\nused_memory:87061960\nused_memory_human:83.03M\nused_memory_rss:91430912\nused_memory_peak:1018693576\nused_memory_peak_human:971.50M\nmem_fragmentation_ratio:1.05\nmem_allocator:jemalloc-2.2.1\nloading:0\naof_enabled:1\nchanges_since_last_save:0\nbgsave_in_progress:0\nlast_save_time:1318870582\nbgrewriteaof_in_progress:0\ntotal_connections_received:2710\ntotal_commands_processed:14067\nexpired_keys:0\nevicted_keys:0\nkeyspace_hits:1023\nkeyspace_misses:8950\npubsub_channels:0\npubsub_patterns:0\nlatest_fork_usec:27831\nvm_enabled:0\nrole:master\naof_current_size:86031166\naof_base_size:0\naof_pending_rewrite:0\ndb0:key\n[2684] 17 Oct 16:56:31 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7fd1abedb215]\n[2684] 17 Oct 16:56:31 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7fd1abedb215]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server() [0x40ca99]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(connectWithMaster+0x1a) [0x41c14a]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(replicationCron+0xe6) [0x41c416]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(serverCron+0x1c8) [0x40fa78]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(aeProcessEvents+0x1db) [0x40c45b]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(aeMain+0x2e) [0x40c70e]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server(main+0x106) [0x40b606]\n[2684] 17 Oct 16:56:31 # /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed) [0x7fd1abe0c30d]\n[2684] 17 Oct 16:56:31 # /usr/local/bin/redis-server() [0x40b72d]\n```\n",
  "state": "closed",
  "created_at": "2011-10-17T17:30:26Z",
  "updated_at": "2014-05-30T23:10:15Z",
  "closed_at": "2011-10-18T17:59:50Z",
  "labels": [
    "critical bug",
    "WAITING-OP-REPLY"
  ],
  "comments_data": [
    {
      "id": 2432506,
      "user": "antirez",
      "created_at": "2011-10-17T18:11:19Z",
      "body": "Hello Jan,\n\nthanks for reporting, this is an \"interesting\" bug I'll try to get fixed within 24 hours. From the stack trace I believe there is some kind of HA daemon issuing \"SLAVE OF NONE\" to your slave instances when the master is unavailable, is it possible?\nI see this from this line:\n\n[2710] 17 Oct 16:56:25 \\* MASTER MODE enabled (user request)\n\nAt the same time however this instance got a request for another slave to use it as master (however this is unrelated to the bug perhaps, need to check this better).\n\nSo there is a bug for sure, but it is not on slave reconnection. Apparently the bug happens once the slave is elected to master. It still tries to reconnect via non blocking connect.\n\nThis is a big bug I'm glad you found it. More news ASAP.\n"
    },
    {
      "id": 2432593,
      "user": "antirez",
      "created_at": "2011-10-17T18:16:59Z",
      "body": "for Jan: please if you can send me more output than this, that is, a few lines more on the top part of the three instances. Thanks!\n"
    },
    {
      "id": 2432638,
      "user": "janoberst",
      "created_at": "2011-10-17T18:20:09Z",
      "body": "I just learned a bit more about this: I believe these crashes occur when a client (A) tries to become slave of an instance (B) that itself is the slave of an instance (C) that just started to become unreachable and that refuses the connection.\n\nIn my case, I issue a \"SLAVEOF NO ONE\" command to instance B shortly before letting A become a slave of B. Judging on the log output that command seems to not be having an effect.\n\nI'm using redis-py to issue these commands quite rapidly, maybe I'm not giving the system enough time to handle all the broken connections?\n"
    },
    {
      "id": 2432689,
      "user": "janoberst",
      "created_at": "2011-10-17T18:23:55Z",
      "body": "Didn't refresh before posting there.. Github should check for new comments every few seconds to avoid problems like these.\n\nYou're right, I have a Python daemon that checks availability of the machines and then issues commands to (A) turn the slave into a master and (B) add a new slave to that machine, so that we have a set of master and slave again.\n\nI'll create a timestamped list of commands issued to each of these machines and send you more of the logfile soon.\n"
    },
    {
      "id": 2433649,
      "user": "janoberst",
      "created_at": "2011-10-17T19:36:52Z",
      "body": "I have three instances:\n\nA (10.230.31.208:10001): Master before restart, starts to become unavailable at some point\nB (10.235.29.62:10006): Slave of A before restart, gets \"slave of no one\", then receives slave request from C and crashes\nC (10.235.30.193:10011): Idle before restart, is told to become slave of B, gets stuck in master_link_down state, because B crashed\n\nAt some point, I restart machine (3) that hosts A. In the end, machine A has been shut down, B will be crashed, and C will be stuck with master_link_down.\n\nstep 1: I stop server 3:\n\n```\n[1139] 17 Oct 18:59:13 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[1139] 17 Oct 18:59:13 - 0 clients connected (1 slaves), 135251576 bytes in use\n[1139] 17 Oct 18:59:13 # Received SIGTERM, scheduling shutdown...\n[1139] 17 Oct 18:59:14 # User requested shutdown...\n[1139] 17 Oct 18:59:14 * Calling fsync() on the AOF file.\n[1139] 17 Oct 18:59:14 # Redis is now ready to exit, bye bye...\n```\n\nstep 2: Daemon notices that server 3 is down:\n\n```\n18:59:26.225980 REDIS ERROR: Error connecting to 10.230.31.208:10001. timed out.\n```\n\nstep 3: Daemon tells B to decouple from master:\n\n```\n18:59:45.660447 ISSUEING 'SLAVEOF NO ONE' for 10.235.29.62:10006 START. Current role: slave.\n18:59:45.662548 ISSUEING 'SLAVEOF NO ONE' for 10.235.29.62:10006 SUCCEEDED. New role: master.\n```\n\nstep 3: Log from B:\n\n```\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45512\n[2818] 17 Oct 18:59:45 * MASTER MODE enabled (user request)\n[2818] 17 Oct 18:59:45 - Client closed connection\n```\n\nstep 4: Daemon does some internal reshuffling, and then tells B the same thing again, although it is already marked as master, just to be sure:\n\n```\n18:59:45.668654 ISSUEING 'SLAVEOF NO ONE' for 10.235.29.62:10006 START. Current role: master.\n18:59:45.671713 ISSUEING 'SLAVEOF NO ONE' for 10.235.29.62:10006 SUCCEEDED. New role: master.\n```\n\nstep 5: Daemon tells C to be \"slaveof no one\", just to be sure it isn't connected to anyone else:\n\n```\n18:59:45.834131 ISSUEING 'SLAVEOF NO ONE' for N[10.235.30.193:10011] START. Current role: master\n18:59:45.836149 ISSUEING 'SLAVEOF NO ONE' for N[10.235.30.193:10011] SUCCEEDED. New role: master\n```\n\nstep 6: Daemon tells C to become slave of B:\n\n```\n18:59:45.839513 ISSUEING 'SLAVE OF 10.235.29.62 10006' to N[10.235.30.193:10011]\n18:59:45.840406 ISSUEING 'SLAVE OF 10.235.29.62 10006' to N[10.235.30.193:10011] SUCCEEDED\n```\n\nstep 7: This is where B starts the sync:\n\n```\n[2818] 17 Oct 18:59:45 * Slave ask for synchronization\n[2818] 17 Oct 18:59:45 * Starting BGSAVE for SYNC\n[2818] 17 Oct 18:59:45 * Background saving started by pid 2985\n[2818] 17 Oct 18:59:49 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:50 - 0 clients connected (1 slaves), 135251792 bytes in use\n[2818] 17 Oct 18:59:55 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:55 - 0 clients connected (1 slaves), 135251792 bytes in use\n```\n\nstep 8: During that sync, C reports link down and has a sync in progress for a few seconds:\n\n```\n18:59:48.849669 10.235.30.193:10011 >> master link status: down / master sync in progress: 1\n[ Omitted: Daemon just reads INFO and prints link status and sync in progress every second ]\n18:59:59.498956 10.235.30.193:10011 >> master link status: down / master sync in progress: 1\n```\n\nstep 9: This is where B crashes:\n\n```\n[2985] 17 Oct 18:59:57 * DB saved on disk\n[2818] 17 Oct 18:59:57 * Background saving terminated with success\n[2818] 17 Oct 18:59:59 * Synchronization with slave succeeded\n[2818] 17 Oct 18:59:59 * Non blocking connect for SYNC fired the event.\n[2818] 17 Oct 18:59:59 # I/O error writing to MASTER: Connection refused\n[2818] 17 Oct 19:00:00 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 19:00:00 - 0 clients connected (1 slaves), 135251792 bytes in use\n[2818] 17 Oct 19:00:00 * Connecting to MASTER...\n[2818] 17 Oct 19:00:00 # ======= Ooops! Redis 2.4.1 got signal: -11- =======\n```\n\nstep 10: At some point \"sync in progress\" changes back to 0, but master link still down:\n\n```\n19:00:00.501939 10.235.30.193:10011 >> master link status: down / master sync in progress: 0\n19:00:01.519828 10.235.30.193:10011 >> master link status: down / master sync in progress: 0\n[...]\n```\n\nThe full log for the crashed instance:\n\n```\n[2818] 17 Oct 18:59:13 - Client closed connection\n[2818] 17 Oct 18:59:13 - Accepted 10.235.23.131:44648\n[2818] 17 Oct 18:59:13 - Client closed connection\n[2818] 17 Oct 18:59:14 - Client closed connection\n[2818] 17 Oct 18:59:14 - Accepted 10.235.23.131:44749\n[2818] 17 Oct 18:59:14 - Client closed connection\n[2818] 17 Oct 18:59:14 - Accepted 10.235.23.131:44750\n[2818] 17 Oct 18:59:14 - Client closed connection\n[2818] 17 Oct 18:59:14 - Accepted 10.235.23.131:44751\n[2818] 17 Oct 18:59:14 - Client closed connection\n[2818] 17 Oct 18:59:14 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:14 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:14 * Connecting to MASTER...\n[2818] 17 Oct 18:59:14 * MASTER <-> SLAVE sync started\n[2818] 17 Oct 18:59:19 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:19 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:24 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:24 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:29 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:29 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:34 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:34 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44877\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44878\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44879\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44880\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44937\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:44938\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45041\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45042\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45043\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45044\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45095\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45096\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45097\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:37 - Accepted 10.235.23.131:45098\n[2818] 17 Oct 18:59:37 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45135\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45327\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45357\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45358\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45359\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45360\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45443\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45444\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45445\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:38 - Accepted 10.235.23.131:45446\n[2818] 17 Oct 18:59:38 - Client closed connection\n[2818] 17 Oct 18:59:39 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:39 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:44 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:44 - 0 clients connected (0 slaves), 135243112 bytes in use\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45509\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45510\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45511\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45512\n[2818] 17 Oct 18:59:45 * MASTER MODE enabled (user request)\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45513\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45514\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45515\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45516\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.23.131:45681\n[2818] 17 Oct 18:59:45 - Client closed connection\n[2818] 17 Oct 18:59:45 - Accepted 10.235.30.193:58045\n[2818] 17 Oct 18:59:45 * Slave ask for synchronization\n[2818] 17 Oct 18:59:45 * Starting BGSAVE for SYNC\n[2818] 17 Oct 18:59:45 * Background saving started by pid 2985\n[2818] 17 Oct 18:59:49 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:50 - 0 clients connected (1 slaves), 135251792 bytes in use\n[2818] 17 Oct 18:59:55 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 18:59:55 - 0 clients connected (1 slaves), 135251792 bytes in use\n[2985] 17 Oct 18:59:57 * DB saved on disk\n[2818] 17 Oct 18:59:57 * Background saving terminated with success\n[2818] 17 Oct 18:59:59 * Synchronization with slave succeeded\n[2818] 17 Oct 18:59:59 * Non blocking connect for SYNC fired the event.\n[2818] 17 Oct 18:59:59 # I/O error writing to MASTER: Connection refused\n[2818] 17 Oct 19:00:00 - DB 0: 134 keys (0 volatile) in 256 slots HT.\n[2818] 17 Oct 19:00:00 - 0 clients connected (1 slaves), 135251792 bytes in use\n[2818] 17 Oct 19:00:00 * Connecting to MASTER...\n[2818] 17 Oct 19:00:00 # ======= Ooops! Redis 2.4.1 got signal: -11- =======\n[2818] 17 Oct 19:00:00 # redis_version:2.4.1\nredis_git_sha1:ab648d97\nredis_git_dirty:0\narch_bits:64\nmultiplexing_api:epoll\nprocess_id:2818\nuptime_in_seconds:316\nuptime_in_days:0\nlru_clock:1864376\nused_cpu_sys:2.97\nused_cpu_user:0.73\nused_cpu_sys_children:0.20\nused_cpu_user_children:0.97\nconnected_clients:0\nconnected_slaves:1\nclient_longest_output_list:0\nclient_biggest_input_buf:0\nblocked_clients:0\nused_memory:135251792\nused_memory_human:128.99M\nused_memory_rss:139513856\nused_memory_peak:137242664\nused_memory_peak_human:130.88M\nmem_fragmentation_ratio:1.03\nmem_allocator:jemalloc-2.2.1\nloading:0\naof_enabled:1\nchanges_since_last_save:0\nbgsave_in_progress:0\nlast_save_time:1318877997\nbgrewriteaof_in_progress:0\ntotal_connections_received:698\ntotal_commands_processed:1546\nexpired_keys:0\nevicted_keys:0\nkeyspace_hits:0\nkeyspace_misses:157\npubsub_channels:0\npubsub_patterns:0\nlatest_fork_usec:37779\nvm_enabled:0\nrole:master\naof_current_size:134078698\naof_base_size:0\naof_pending_rewrite:0\ndb0:keys=134\n[2818] 17 Oct 19:00:00 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f8a39989215]\n[2818] 17 Oct 19:00:00 # /lib/x86_64-linux-gnu/libc.so.6(inet_aton+0x35) [0x7f8a39989215]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server() [0x40ca99]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(connectWithMaster+0x1a) [0x41c14a]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(replicationCron+0xe6) [0x41c416]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(serverCron+0x1c8) [0x40fa78]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(aeProcessEvents+0x1db) [0x40c45b]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(aeMain+0x2e) [0x40c70e]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server(main+0x106) [0x40b606]\n[2818] 17 Oct 19:00:00 # /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xed) [0x7f8a398ba30d]\n[2818] 17 Oct 19:00:00 # /usr/local/bin/redis-server() [0x40b72d]\n```\n"
    },
    {
      "id": 2433895,
      "user": "antirez",
      "created_at": "2011-10-17T19:56:17Z",
      "body": "Jan: just to say you that his kind of help is really appreciated :)\nTomorrow morning this is on top on my TODO list, I hope to get this fixed with the informations you already provided, otherwise I'll use your kindness again to ask more questions.\n\nThanks!\nSalvatore\n"
    },
    {
      "id": 2439541,
      "user": "antirez",
      "created_at": "2011-10-18T09:12:17Z",
      "body": "Hello Jan, I was not able to reproduce the issue, but following the logic of what happened I think I fixed the issue in commit \"98f5abb\".\n\nWhat was happening is that the event to connect with the master in a non blocking way fired _after_ we already switched the instance into a slave. That even failed to accomplish its work, but as a side effect turned again the replication status into \"connecting\". So at the next cron loop the instance tried again to connect with the master. Since the instance is not a slave now the master host filed was set to NULL, causing a crash.\n\nPlease could you verify that latest commit in 2.4 branch fixes it?\n\nThanks again as only the detailed description made this fix easy (assuming it's fixed ;) since I was not able to reproduce.\n"
    },
    {
      "id": 2442444,
      "user": "janoberst",
      "created_at": "2011-10-18T14:37:10Z",
      "body": "Hi Salvatore,\n\nThank you so much for getting to this so quickly. Your change definitely fixed the server crash. I re-ran my test twice and both times all 48 redis processes survived. Awesome.\n\nBut maybe this just made way for us to see the actual problem. Here's this morning's log. 10.235.27.6 is my Daemon, all the other IPs have changed, too. But the setup is the same. B is 10.230.31.186:10005, and C is 10.230.17.150:10010.\n\nInstance C is still trying to become slave of instance B, but doesn't succeed because of \"Error allocating resoures for the client\". This is a rather big DB with around 400MB, but the same things happen at around 40MB, too. The objects are seriously huge at around 1MB per object. Maybe that has something to do with it? If you want to, I can share the dataset with you, it's just random strings, lists, sets, sorted sets and hashes of 1MB length. The error never occurred on datasets with smaller objects.\n\nAfter sending \"SLAVE OF NO ONE\" to B and setting up C as slave of B, here's the log of B:\n\n```\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57487\n[2786] 18 Oct 14:18:17 * MASTER MODE enabled (user request)\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57488\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57489\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57490\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57491\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.235.27.6:57656\n[2786] 18 Oct 14:18:17 - Client closed connection\n[2786] 18 Oct 14:18:17 - Accepted 10.230.17.150:34267\n[2786] 18 Oct 14:18:17 # Error allocating resoures for the client\n[2786] 18 Oct 14:18:18 - DB 0: 411 keys (0 volatile) in 512 slots HT.\n[2786] 18 Oct 14:18:18 - 0 clients connected (0 slaves), 412321256 bytes in use\n[2786] 18 Oct 14:18:18 - Accepted 10.230.17.150:34269\n[2786] 18 Oct 14:18:18 # Error allocating resoures for the client\n[2786] 18 Oct 14:18:19 - Accepted 10.230.17.150:34271\n[2786] 18 Oct 14:18:19 # Error allocating resoures for the client\n[2786] 18 Oct 14:18:20 - Accepted 10.230.17.150:34273\n[2786] 18 Oct 14:18:20 # Error allocating resoures for the client\n```\n\nlog for C:\n\n```\n[2854] 18 Oct 14:18:17 - Accepted 10.235.27.6:55257\n[2854] 18 Oct 14:18:17 * SLAVE OF 10.230.31.186:10005 enabled (user request)\n[2854] 18 Oct 14:18:17 - Client closed connection\n[2854] 18 Oct 14:18:17 - Accepted 10.235.27.6:55258\n[2854] 18 Oct 14:18:17 - Client closed connection\n[2854] 18 Oct 14:18:18 * Connecting to MASTER...\n[2854] 18 Oct 14:18:18 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:18 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:18 # I/O error reading bulk count from MASTER: Connection reset by peer\n[2854] 18 Oct 14:18:18 - Accepted 10.235.27.6:55259\n[2854] 18 Oct 14:18:18 - Client closed connection\n[2854] 18 Oct 14:18:19 - 0 clients connected (0 slaves), 717592 bytes in use\n[2854] 18 Oct 14:18:19 * Connecting to MASTER...\n[2854] 18 Oct 14:18:19 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:19 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:19 # I/O error reading bulk count from MASTER: Operation now in progress\n[2854] 18 Oct 14:18:19 - Accepted 10.235.27.6:55260\n[2854] 18 Oct 14:18:19 - Client closed connection\n[2854] 18 Oct 14:18:20 * Connecting to MASTER...\n[2854] 18 Oct 14:18:20 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:20 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:20 # I/O error reading bulk count from MASTER: Connection reset by peer\n[2854] 18 Oct 14:18:20 - Accepted 10.235.27.6:55268\n[2854] 18 Oct 14:18:20 - Client closed connection\n[2854] 18 Oct 14:18:21 * Connecting to MASTER...\n[2854] 18 Oct 14:18:21 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:21 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:21 # I/O error reading bulk count from MASTER: Connection reset by peer\n[2854] 18 Oct 14:18:21 - Accepted 10.235.27.6:55269\n[2854] 18 Oct 14:18:21 - Client closed connection\n[2854] 18 Oct 14:18:22 * Connecting to MASTER...\n[2854] 18 Oct 14:18:22 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:22 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:22 # I/O error reading bulk count from MASTER: Connection reset by peer\n[2854] 18 Oct 14:18:22 - Accepted 10.235.27.6:55270\n[2854] 18 Oct 14:18:22 - Client closed connection\n[2854] 18 Oct 14:18:23 * Connecting to MASTER...\n[2854] 18 Oct 14:18:23 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:23 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:23 # I/O error reading bulk count from MASTER: Operation now in progress\n[2854] 18 Oct 14:18:23 - Accepted 10.235.27.6:55271\n[2854] 18 Oct 14:18:23 - Client closed connection\n[2854] 18 Oct 14:18:24 - 0 clients connected (0 slaves), 717592 bytes in use\n[2854] 18 Oct 14:18:24 * Connecting to MASTER...\n[2854] 18 Oct 14:18:24 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:24 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:24 # I/O error reading bulk count from MASTER: Connection reset by peer\n[2854] 18 Oct 14:18:24 - Accepted 10.235.27.6:55272\n[2854] 18 Oct 14:18:24 - Client closed connection\n[2854] 18 Oct 14:18:25 * Connecting to MASTER...\n[2854] 18 Oct 14:18:25 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:25 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:25 # I/O error reading bulk count from MASTER: Operation now in progress\n[2854] 18 Oct 14:18:25 - Accepted 10.235.27.6:55273\n[2854] 18 Oct 14:18:25 - Client closed connection\n[2854] 18 Oct 14:18:26 * Connecting to MASTER...\n[2854] 18 Oct 14:18:26 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:26 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:26 # I/O error reading bulk count from MASTER: Operation now in progress\n[2854] 18 Oct 14:18:26 - Accepted 10.235.27.6:55274\n[2854] 18 Oct 14:18:26 - Client closed connection\n[2854] 18 Oct 14:18:27 * Connecting to MASTER...\n[2854] 18 Oct 14:18:27 * MASTER <-> SLAVE sync started\n[2854] 18 Oct 14:18:27 * Non blocking connect for SYNC fired the event.\n[2854] 18 Oct 14:18:27 # I/O error reading bulk count from MASTER: Operation now in progress\n```\n"
    },
    {
      "id": 2442571,
      "user": "antirez",
      "created_at": "2011-10-18T14:47:32Z",
      "body": "Hello Jan, thanks for the quick reply!\n\nThe problem we have here is apparently not related to our previous problem, and it is a bit strange, as this error message can only be the result of the following lines of code failing:\n\n```\n    if (aeCreateFileEvent(server.el,fd,AE_READABLE,\n        readQueryFromClient, c) == AE_ERR)\n    {\n        close(fd);\n        zfree(c);\n        return NULL;\n    }\n```\n\nit is not possible it's an out of memory so the only possibility is that we fail trying to register the event in the epool interface.\n\nSo I'm already in need for more debugging / information, specifically the INFO output of the instance logging \"# Error allocating resoures for the client\" can be helpful, and even more, and \"strace -p <pid>\" trace while the instance is outputting this error.\n\nThanks again,\nSalvatore\n"
    },
    {
      "id": 2442690,
      "user": "antirez",
      "created_at": "2011-10-18T14:57:17Z",
      "body": "Hello Jan, looks like this are the logs related to \"C\" perhaps? I need to look at logs from \"B\" since actually this is the instance that is misbehaving, \"C\" bad logs are only the result of \"B\" not being very friendly ;)\n\nSalvatore\n"
    },
    {
      "id": 2442707,
      "user": "janoberst",
      "created_at": "2011-10-18T14:58:58Z",
      "body": "Hi Salvatore,\n\nHere's the logs you were looking for. I mixed up the two instances before. Sorry :)\n\nINFO for the failing machine:\n\n```\nubuntu@ip-10-230-31-186:/mnt/logs$ redis-cli -p 10005\nredis 127.0.0.1:10005> info\nError: Connection reset by peer\n```\n\nstrace of the failing machine:\n\n```\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949819, 968284}, NULL) = 0\ngettimeofday({1318949819, 968394}, NULL) = 0\ngettimeofday({1318949819, 968520}, NULL) = 0\ngettimeofday({1318949819, 968667}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949820, 69078}, NULL) = 0\ngettimeofday({1318949820, 69215}, NULL) = 0\ngettimeofday({1318949820, 69320}, NULL) = 0\ngettimeofday({1318949820, 69444}, NULL) = 0\nepoll_wait(3, {{EPOLLIN, {u32=4, u64=4}}}, 10240, 100) = 1\naccept(4, {sa_family=AF_INET, sin_port=htons(36536), sin_addr=inet_addr(\"10.230.17.150\")}, [16]) = 5\nopen(\"/mnt/logs/instance_5.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 6\nfstat(6, {st_mode=S_IFREG|0644, st_size=427380, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff45fbff000\nfstat(6, {st_mode=S_IFREG|0644, st_size=427380, ...}) = 0\nlseek(6, 427380, SEEK_SET)              = 427380\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(6, \"[2786] 18 Oct 14:57:00 - Accepte\"..., 54) = 54\nclose(6)                                = 0\nmunmap(0x7ff45fbff000, 4096)            = 0\nfcntl(5, F_GETFL)                       = 0x2 (flags O_RDWR)\nfcntl(5, F_SETFL, O_RDWR|O_NONBLOCK)    = 0\nsetsockopt(5, SOL_TCP, TCP_NODELAY, [1], 4) = 0\nepoll_ctl(3, EPOLL_CTL_MOD, 5, {EPOLLIN|EPOLLOUT, {u32=5, u64=5}}) = -1 ENOENT (No such file or directory)\nclose(5)                                = 0\nopen(\"/mnt/logs/instance_5.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 5\nfstat(5, {st_mode=S_IFREG|0644, st_size=427434, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff45fbff000\nfstat(5, {st_mode=S_IFREG|0644, st_size=427434, ...}) = 0\nlseek(5, 427434, SEEK_SET)              = 427434\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(5, \"[2786] 18 Oct 14:57:00 # Error a\"..., 66) = 66\nclose(5)                                = 0\nmunmap(0x7ff45fbff000, 4096)            = 0\nclose(5)                                = -1 EBADF (Bad file descriptor)\ngettimeofday({1318949820, 154115}, NULL) = 0\ngettimeofday({1318949820, 154324}, NULL) = 0\nepoll_wait(3, {}, 10240, 15)            = 0\ngettimeofday({1318949820, 169859}, NULL) = 0\ngettimeofday({1318949820, 170018}, NULL) = 0\ngettimeofday({1318949820, 170164}, NULL) = 0\ngettimeofday({1318949820, 170303}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949820, 270833}, NULL) = 0\ngettimeofday({1318949820, 271008}, NULL) = 0\ngettimeofday({1318949820, 271167}, NULL) = 0\ngettimeofday({1318949820, 271301}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\n```\n\nhere's INFO for the requesting instance:\n\n```\nredis 127.0.0.1:10010> info\nredis_version:2.4.1\nredis_git_sha1:1270a136\nredis_git_dirty:0\narch_bits:64\nmultiplexing_api:epoll\nprocess_id:2854\nuptime_in_seconds:2433\nuptime_in_days:0\nlru_clock:1871523\nused_cpu_sys:1.72\nused_cpu_user:0.71\nused_cpu_sys_children:0.00\nused_cpu_user_children:0.00\nconnected_clients:1\nconnected_slaves:0\nclient_longest_output_list:0\nclient_biggest_input_buf:0\nblocked_clients:0\nused_memory:726192\nused_memory_human:709.17K\nused_memory_rss:1892352\nused_memory_peak:726192\nused_memory_peak_human:709.17K\nmem_fragmentation_ratio:2.61\nmem_allocator:jemalloc-2.2.1\nloading:0\naof_enabled:1\nchanges_since_last_save:2\nbgsave_in_progress:0\nlast_save_time:1318947041\nbgrewriteaof_in_progress:0\ntotal_connections_received:2296\ntotal_commands_processed:4590\nexpired_keys:0\nevicted_keys:0\nkeyspace_hits:0\nkeyspace_misses:0\npubsub_channels:0\npubsub_patterns:0\nlatest_fork_usec:499\nvm_enabled:0\nrole:slave\naof_current_size:41\naof_base_size:0\naof_pending_rewrite:0\nmaster_host:10.230.31.186\nmaster_port:10005\nmaster_link_status:down\nmaster_last_io_seconds_ago:-1\nmaster_sync_in_progress:0\nmaster_link_down_since_seconds:1318949475\n```\n\nstrace for the requesting instance:\n\n```\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949585, 604891}, NULL) = 0\ngettimeofday({1318949585, 604986}, NULL) = 0\ngettimeofday({1318949585, 605076}, NULL) = 0\ngettimeofday({1318949585, 605169}, NULL) = 0\nepoll_wait(3, {}, 10240, 99)            = 0\ngettimeofday({1318949585, 704492}, NULL) = 0\ngettimeofday({1318949585, 704585}, NULL) = 0\ngettimeofday({1318949585, 704672}, NULL) = 0\ngettimeofday({1318949585, 704765}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949585, 805101}, NULL) = 0\ngettimeofday({1318949585, 805195}, NULL) = 0\ngettimeofday({1318949585, 805284}, NULL) = 0\ngettimeofday({1318949585, 805378}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949585, 905730}, NULL) = 0\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 5\nfstat(5, {st_mode=S_IFREG|0644, st_size=818750, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(5, {st_mode=S_IFREG|0644, st_size=818750, ...}) = 0\nlseek(5, 818750, SEEK_SET)              = 818750\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(5, \"[2854] 18 Oct 14:53:05 * Connect\"..., 49) = 49\nclose(5)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\nsocket(PF_INET, SOCK_STREAM, IPPROTO_IP) = 5\nsetsockopt(5, SOL_SOCKET, SO_REUSEADDR, [1], 4) = 0\nfcntl(5, F_GETFL)                       = 0x2 (flags O_RDWR)\nfcntl(5, F_SETFL, O_RDWR|O_NONBLOCK)    = 0\nconnect(5, {sa_family=AF_INET, sin_port=htons(10005), sin_addr=inet_addr(\"10.230.31.186\")}, 16) = -1 EINPROGRESS (Operation now in progress)\nepoll_ctl(3, EPOLL_CTL_ADD, 5, {EPOLLIN|EPOLLOUT, {u32=5, u64=5}}) = 0\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 7\nfstat(7, {st_mode=S_IFREG|0644, st_size=818799, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(7, {st_mode=S_IFREG|0644, st_size=818799, ...}) = 0\nlseek(7, 818799, SEEK_SET)              = 818799\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(7, \"[2854] 18 Oct 14:53:05 * MASTER \"..., 55) = 55\nclose(7)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\ngettimeofday({1318949585, 908245}, NULL) = 0\ngettimeofday({1318949585, 908348}, NULL) = 0\ngettimeofday({1318949585, 908427}, NULL) = 0\nepoll_wait(3, {{EPOLLIN|EPOLLOUT, {u32=5, u64=5}}}, 10240, 100) = 1\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 7\nfstat(7, {st_mode=S_IFREG|0644, st_size=818854, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(7, {st_mode=S_IFREG|0644, st_size=818854, ...}) = 0\nlseek(7, 818854, SEEK_SET)              = 818854\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(7, \"[2854] 18 Oct 14:53:05 * Non blo\"..., 72) = 72\nclose(7)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\nepoll_ctl(3, EPOLL_CTL_DEL, 5, {0, {u32=5, u64=5}}) = 0\nselect(6, [], [5], [], {1, 0})          = 1 (out [5], left {0, 999998})\nwrite(5, \"SYNC \\r\\n\", 7)                = 7\nopen(\"temp-1318949585.2854.rdb\", O_WRONLY|O_CREAT|O_EXCL, 0644) = 7\nepoll_ctl(3, EPOLL_CTL_ADD, 5, {EPOLLIN, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 909917}, NULL) = 0\ngettimeofday({1318949585, 909995}, NULL) = 0\nepoll_wait(3, {{EPOLLIN|EPOLLERR|EPOLLHUP, {u32=5, u64=5}}}, 10240, 99) = 1\nselect(6, [5], [], [], {1, 0})          = 1 (in [5], left {0, 999998})\nread(5, \"\", 1)                          = 0\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 8\nfstat(8, {st_mode=S_IFREG|0644, st_size=818926, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(8, {st_mode=S_IFREG|0644, st_size=818926, ...}) = 0\nlseek(8, 818926, SEEK_SET)              = 818926\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(8, \"[2854] 18 Oct 14:53:05 # I/O err\"..., 93) = 93\nclose(8)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\nepoll_ctl(3, EPOLL_CTL_DEL, 5, {0, {u32=5, u64=5}}) = 0\nclose(5)                                = 0\nclose(7)                                = 0\nunlink(\"temp-1318949585.2854.rdb\")      = 0\ngettimeofday({1318949585, 911476}, NULL) = 0\ngettimeofday({1318949585, 911556}, NULL) = 0\nepoll_wait(3, {{EPOLLIN, {u32=4, u64=4}}}, 10240, 97) = 1\naccept(4, {sa_family=AF_INET, sin_port=htons(59031), sin_addr=inet_addr(\"10.235.27.6\")}, [16]) = 5\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 7\nfstat(7, {st_mode=S_IFREG|0644, st_size=819019, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(7, {st_mode=S_IFREG|0644, st_size=819019, ...}) = 0\nlseek(7, 819019, SEEK_SET)              = 819019\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(7, \"[2854] 18 Oct 14:53:05 - Accepte\"..., 52) = 52\nclose(7)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\nfcntl(5, F_GETFL)                       = 0x2 (flags O_RDWR)\nfcntl(5, F_SETFL, O_RDWR|O_NONBLOCK)    = 0\nsetsockopt(5, SOL_TCP, TCP_NODELAY, [1], 4) = 0\nepoll_ctl(3, EPOLL_CTL_ADD, 5, {EPOLLIN, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 913381}, NULL) = 0\ngettimeofday({1318949585, 913464}, NULL) = 0\nepoll_wait(3, {{EPOLLIN, {u32=5, u64=5}}}, 10240, 95) = 1\nread(5, \"*1\\r\\n$4\\r\\nPING\\r\\n\", 1024)   = 14\ngettimeofday({1318949585, 913722}, NULL) = 0\nepoll_ctl(3, EPOLL_CTL_MOD, 5, {EPOLLIN|EPOLLOUT, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 913883}, NULL) = 0\ngettimeofday({1318949585, 913962}, NULL) = 0\ngettimeofday({1318949585, 914038}, NULL) = 0\nepoll_wait(3, {{EPOLLOUT, {u32=5, u64=5}}}, 10240, 94) = 1\nwrite(5, \"+PONG\\r\\n\", 7)                = 7\nepoll_ctl(3, EPOLL_CTL_MOD, 5, {EPOLLIN, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 914383}, NULL) = 0\ngettimeofday({1318949585, 914464}, NULL) = 0\nepoll_wait(3, {{EPOLLIN, {u32=5, u64=5}}}, 10240, 94) = 1\nread(5, \"*1\\r\\n$4\\r\\nINFO\\r\\n\", 1024)   = 14\ngettimeofday({1318949585, 914774}, NULL) = 0\ngetrusage(RUSAGE_SELF, {ru_utime={0, 764047}, ru_stime={1, 828114}, ...}) = 0\ngetrusage(RUSAGE_CHILDREN, {ru_utime={0, 0}, ru_stime={0, 0}, ...}) = 0\nopen(\"/proc/2854/stat\", O_RDONLY)       = 7\nread(7, \"2854 (redis-server) R 2852 2852 \"..., 4096) = 245\nclose(7)                                = 0\nopen(\"/proc/2854/stat\", O_RDONLY)       = 7\nread(7, \"2854 (redis-server) R 2852 2852 \"..., 4096) = 245\nclose(7)                                = 0\nepoll_ctl(3, EPOLL_CTL_MOD, 5, {EPOLLIN|EPOLLOUT, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 915778}, NULL) = 0\ngettimeofday({1318949585, 915860}, NULL) = 0\ngettimeofday({1318949585, 915939}, NULL) = 0\nepoll_wait(3, {{EPOLLOUT, {u32=5, u64=5}}}, 10240, 93) = 1\nwrite(5, \"$1165\\r\\nredis_version:2.4.1\\r\\nredi\"..., 1174) = 1174\nepoll_ctl(3, EPOLL_CTL_MOD, 5, {EPOLLIN, {u32=5, u64=5}}) = 0\ngettimeofday({1318949585, 916302}, NULL) = 0\ngettimeofday({1318949585, 916381}, NULL) = 0\nepoll_wait(3, {{EPOLLIN, {u32=5, u64=5}}}, 10240, 92) = 1\nread(5, \"\", 1024)                       = 0\nopen(\"/mnt/logs/instance_10.log\", O_WRONLY|O_CREAT|O_APPEND, 0666) = 7\nfstat(7, {st_mode=S_IFREG|0644, st_size=819071, ...}) = 0\nmmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f60e72e9000\nfstat(7, {st_mode=S_IFREG|0644, st_size=819071, ...}) = 0\nlseek(7, 819071, SEEK_SET)              = 819071\nstat(\"/etc/localtime\", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0\nwrite(7, \"[2854] 18 Oct 14:53:05 - Client \"..., 50) = 50\nclose(7)                                = 0\nmunmap(0x7f60e72e9000, 4096)            = 0\nepoll_ctl(3, EPOLL_CTL_DEL, 5, {0, {u32=5, u64=5}}) = 0\nclose(5)                                = 0\ngettimeofday({1318949585, 917960}, NULL) = 0\ngettimeofday({1318949585, 918038}, NULL) = 0\nepoll_wait(3, {}, 10240, 90)            = 0\ngettimeofday({1318949586, 8400}, NULL)  = 0\ngettimeofday({1318949586, 8546}, NULL)  = 0\ngettimeofday({1318949586, 8662}, NULL)  = 0\ngettimeofday({1318949586, 8818}, NULL)  = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949586, 109251}, NULL) = 0\ngettimeofday({1318949586, 109367}, NULL) = 0\ngettimeofday({1318949586, 109501}, NULL) = 0\ngettimeofday({1318949586, 109643}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949586, 210076}, NULL) = 0\ngettimeofday({1318949586, 210195}, NULL) = 0\ngettimeofday({1318949586, 210326}, NULL) = 0\ngettimeofday({1318949586, 210467}, NULL) = 0\nepoll_wait(3, {}, 10240, 100)           = 0\ngettimeofday({1318949586, 310827}, NULL) = 0\ngettimeofday({1318949586, 310926}, NULL) = 0\ngettimeofday({1318949586, 311015}, NULL) = 0\ngettimeofday({1318949586, 311103}, NULL) = 0\nepoll_wait(3, {}, 10240, 99)            = 0\n```\n"
    },
    {
      "id": 2442896,
      "user": "antirez",
      "created_at": "2011-10-18T15:10:11Z",
      "body": "Great, I think I know what the problem is... just a moment. It is actually that my former fix was not complete, but it is the same bug indeed ;)\n"
    },
    {
      "id": 2442929,
      "user": "antirez",
      "created_at": "2011-10-18T15:12:35Z",
      "body": "Fixed, hopefully latest 2.4 will be fine :) Thanks.\n"
    },
    {
      "id": 2442993,
      "user": "janoberst",
      "created_at": "2011-10-18T15:16:38Z",
      "body": "Sounds great. Thanks for being the greatest dev any open source software has ever had :D\n\nBut enough with the pandering.. I have a dentist appointment now, but will verify asap.\n"
    },
    {
      "id": 2443131,
      "user": "antirez",
      "created_at": "2011-10-18T15:26:31Z",
      "body": "Thanks a lot Jan! It was easy to fix this issue only thanks to your great support and good luck with the dentist! ;)\n"
    },
    {
      "id": 2445093,
      "user": "janoberst",
      "created_at": "2011-10-18T17:53:28Z",
      "body": "I ran my test again, and everything runs smoothly. The old slave becomes the new master, and the new slave properly connects and syncs all data.\n\nGreat. Thank you, Salvatore!\n"
    },
    {
      "id": 2445179,
      "user": "antirez",
      "created_at": "2011-10-18T17:59:50Z",
      "body": "We won! :) Thanks, closing for now in the hope everything will go smooth even in the next days ;) Otherwise I'm here.\n"
    }
  ]
}